data/transactions/verify/txn.go

```diff
@@ -536,3 +557,391 @@ func (w *worksetBuilder) next() (txnGroups [][]transactions.SignedTxn) {
+		return nil, errors.New("MakeStreamVerifier: Could not get header for previous block")
+	}
+
+	nbw := MakeNewBlockWatcher(latestHdr)
+	ledger.RegisterBlockListeners([]ledgercore.BlockListener{nbw})
+
+	return &StreamVerifier{
+		resultChan:       resultChan,
+		stxnChan:         stxnChan,
+		droppedChan:      droppedChan,
+		verificationPool: verificationPool,
+		cache:            cache,
+		nbw:              nbw,
+		ledger:           ledger,
+	}, nil
+}
+
+// Start is called when the verifier is created and whenever it needs to restart after
+// the ctx is canceled
+func (sv *StreamVerifier) Start(ctx context.Context) {
+	sv.ctx = ctx
+	sv.activeLoopWg.Add(1)
+	go sv.batchingLoop()
+}
+
+// WaitForStop waits until the batching loop terminates afer the ctx is cancled
+func (sv *StreamVerifier) WaitForStop() {
+	sv.activeLoopWg.Wait()
+}
+
+func (sv *StreamVerifier) cleanup(pending []*UnverifiedElement) {
+	// report an error for the unchecked txns
+	// drop the messages without reporting if the receiver does not consume
+	for _, uel := range pending {
+		sv.sendResult(uel.TxnGroup, uel.BacklogMessage, errShuttingDownError)
+	}
+}
+
+func (sv *StreamVerifier) batchingLoop() {
+	defer sv.activeLoopWg.Done()
+	timer := time.NewTicker(waitForNextTxnDuration)
+	defer timer.Stop()
+	var added bool
+	var numberOfSigsInCurrent uint64
+	var numberOfTimerResets uint64
+	ue := make([]*UnverifiedElement, 0, 8)
+	defer func() { sv.cleanup(ue) }()
+	for {
+		select {
+		case stx := <-sv.stxnChan:
+			numberOfBatchableSigsInGroup, err := getNumberOfBatchableSigsInGroup(stx.TxnGroup)
+			if err != nil {
+				// wrong number of signatures
+				sv.sendResult(stx.TxnGroup, stx.BacklogMessage, err)
+				continue
+			}
+
+			// if no batchable signatures here, send this as a task of its own
+			if numberOfBatchableSigsInGroup == 0 {
+				err := sv.addVerificationTaskToThePoolNow([]*UnverifiedElement{stx})
+				if err != nil {
+					return
+				}
+				continue // stx is handled, continue
+			}
+
+			// add this txngrp to the list of batchable txn groups
+			numberOfSigsInCurrent = numberOfSigsInCurrent + numberOfBatchableSigsInGroup
+			ue = append(ue, stx)
+			if numberOfSigsInCurrent > txnPerWorksetThreshold {
+				// enough transaction in the batch to efficiently verify
+
+				if numberOfSigsInCurrent > batchSizeBlockLimit {
+					// do not consider adding more txns to this batch.
+					// bypass the exec pool situation and queue anyway
+					// this is to prevent creation of very large batches
+					err := sv.addVerificationTaskToThePoolNow(ue)
+					if err != nil {
+						return
+					}
+					added = true
+				} else {
+					added, err = sv.tryAddVerificationTaskToThePool(ue)
+					if err != nil {
+						return
+					}
+				}
+				if added {
+					numberOfSigsInCurrent = 0
+					ue = make([]*UnverifiedElement, 0, 8)
+					numberOfTimerResets = 0
+				} else {
+					// was not added because of the exec pool buffer length
+					numberOfTimerResets++
```

Comment:
`numberOfTimerResets` is misleading, it is incremented when the exec pool is full, and not related to the timer, it appears to be a safety measure against `l == c` check in the `tryAddVerificationTaskToThePool` function, since the actual backlog queue length could be changed before reading it's capacity.

Reply:
renaming this from numberOfTimerResets to numberOfBatchAttempts